### PR TITLE
Update MinGW details in the README

### LINKED ISSUE #29649: Build failure with MinGW-w64 gcc 5.2.0
## Updated description

We need to document that the standard MSYS2 package for gcc **will not work** when building Rust, you have to install a custom toolchain with a known good version and the win32 threading model, not the pthread threading model.

## Original report

Looks related to https://github.com/rust-lang/rust/pull/29177, perhaps the `gcc_eh` library picked up new dependencies in 5.2.0?

cc @vadimcn 

```
error: linking with `gcc` failed: exit code: 1
note: "gcc" "-Wl,--enable-long-section-names" "-fno-use-linker-plugin" "-Wl,--nxcompat" "-nostdlib" "-Wl,--large-address-aware" "C:\\msys64\\home\\ale\\rust2\\i686-pc-windows-gnu\\stage1\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\dllcrt2.o" "C:\\msys64\\home\\ale\\rust2\\i686-pc-windows-gnu\\stage1\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\rsbegin.o" "-L" "C:\\msys64\\home\\ale\\rust2\\i686-pc-windows-gnu\\stage1\\lib\\rustlib\\i686-pc-windows-gnu\\lib" "i686-pc-windows-gnu/stage1/lib/rustlib/i686-pc-windows-gnu/lib\\std-71b07a99.0.o" "-o" "i686-pc-windows-gnu/stage1/lib/rustlib/i686-pc-windows-gnu/lib\\std-71b07a99.dll" "i686-pc-windows-gnu/stage1/lib/rustlib/i686-pc-windows-gnu/lib\\std-71b07a99.metadata.o" "-nodefaultlibs" "-L" "i686-pc-windows-gnu/rt" "-L" "C:\\msys64\\home\\ale\\rust2\\i686-pc-windows-gnu\\llvm\\Release/lib" "-L" "C:\\msys64\\home\\ale\\rust2\\i686-pc-windows-gnu\\stage1\\lib\\rustlib\\i686-pc-windows-gnu\\lib" "-L" "C:\\msys64\\home\\ale\\rust2\\.rust\\lib\\i686-pc-windows-gnu" "-L" "C:\\msys64\\home\\ale\\rust2\\lib\\i686-pc-windows-gnu" "-Wl,-Bstatic" "-Wl,--whole-archive" "-l" "backtrace" "-Wl,--no-whole-archive" "-Wl,-Bdynamic" "-l" "gcc_eh" "-l" "gcc_eh" "-l" "ws2_32" "-l" "userenv" "-l" "shell32" "-l" "advapi32" "-Wl,--whole-archive" "C:\\msys64\\tmp\\rustc.hZLinbW19bbD\\libcollections-71b07a99.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "C:\\msys64\\tmp\\rustc.hZLinbW19bbD\\liballoc-71b07a99.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "C:\\msys64\\tmp\\rustc.hZLinbW19bbD\\librustc_unicode-71b07a99.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "C:\\msys64\\tmp\\rustc.hZLinbW19bbD\\liballoc_jemalloc-71b07a99.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "C:\\msys64\\tmp\\rustc.hZLinbW19bbD\\liblibc-71b07a99.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "C:\\msys64\\tmp\\rustc.hZLinbW19bbD\\librand-71b07a99.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "C:\\msys64\\tmp\\rustc.hZLinbW19bbD\\libcore-71b07a99.rlib" "-Wl,--no-whole-archive" "-shared" "-l" "compiler-rt" "-lmingwex" "-lmingw32" "-lgcc" "-lmsvcrt" "-luser32" "-lkernel32" "C:\\msys64\\home\\ale\\rust2\\i686-pc-windows-gnu\\stage1\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\rsend.o"
note: C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/5.2.0/libgcc_eh.a(unwind-dw2.o):(.text+0x1a4b): undefined reference to `pthread_once'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/5.2.0/libgcc_eh.a(unwind-dw2-fde.o):(.text+0x1a3): undefined reference to `pthread_mutex_init'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/5.2.0/libgcc_eh.a(unwind-dw2-fde.o):(.text+0xf99): undefined reference to `pthread_once'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/5.2.0/libgcc_eh.a(unwind-dw2-fde.o):(.text+0xfa5): undefined reference to `pthread_mutex_lock'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/5.2.0/libgcc_eh.a(unwind-dw2-fde.o):(.text+0x10b4): undefined reference to `pthread_once'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/5.2.0/libgcc_eh.a(unwind-dw2-fde.o):(.text+0x10c0): undefined reference to `pthread_mutex_lock'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/5.2.0/libgcc_eh.a(unwind-dw2-fde.o):(.text+0x118b): undefined reference to `pthread_once'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/5.2.0/libgcc_eh.a(unwind-dw2-fde.o):(.text+0x1197): undefined reference to `pthread_mutex_lock'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/5.2.0/libgcc_eh.a(unwind-dw2-fde.o):(.text+0x11f9): undefined reference to `pthread_mutex_unlock'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/5.2.0/libgcc_eh.a(unwind-dw2-fde.o):(.text+0x1228): undefined reference to `pthread_mutex_unlock'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/5.2.0/libgcc_eh.a(unwind-dw2-fde.o):(.text+0x12af): undefined reference to `pthread_once'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/5.2.0/libgcc_eh.a(unwind-dw2-fde.o):(.text+0x12bb): undefined reference to `pthread_mutex_lock'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/5.2.0/libgcc_eh.a(unwind-dw2-fde.o):(.text+0x12f7): undefined reference to `pthread_mutex_unlock'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/5.2.0/libgcc_eh.a(unwind-dw2-fde.o):(.text+0x13b6): undefined reference to `pthread_mutex_unlock'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/5.2.0/libgcc_eh.a(unwind-dw2-fde.o):(.text+0xfc4): undefined reference to `pthread_mutex_unlock'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/5.2.0/libgcc_eh.a(unwind-dw2-fde.o):(.text+0x10df): undefined reference to `pthread_mutex_unlock'
collect2.exe: error: ld returned 1 exit status

error: aborting due to previous error

This application has requested the Runtime to terminate it in an unusual way.
Please contact the application's support team for more information.
/home/ale/rust2/mk/target.mk:231: recipe for target 'i686-pc-windows-gnu/stage1/lib/rustlib/i686-pc-windows-gnu/lib/stamp.std' failed
make: *** [i686-pc-windows-gnu/stage1/lib/rustlib/i686-pc-windows-gnu/lib/stamp.std] Error 3
```

### PR DESCRIPTION
**Pull request by steveklabnik**
Fixes #29649 

r? @retep998 @alexcrichton

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -88,7 +89,7 @@ Read ["Installing Rust"] from [The Book].
 > resulting in a segmentation fault when building Rust. In order to avoid this
 > it may be necessary to obtain an earlier version of gcc such as 4.9.x.  
 > Msys's `pacman` will install the latest version, so for the time being it is
-> recommended to skip gcc toolchain installation step above and use [Mingw-Builds]
+> recommended to skip gcc toolchain installation step above and use [Mingw-Builds
```
Comment (rkruppe) [inline README.md]: This is probably not intended.

Comment (steveklabnik) [inline README.md]: whoah, thanks. No idea how that happened.

**Comment by retep998** [comment]
The link on that page takes me to a sourceforge download. I can't say I'm too happy about sourceforge (ublock blocks it saying it is badware), but then again, msys2 has to be obtained through sourceforge as well.

Also, a bit further down, the readme already describes how to obtain an alternate mingw-w64 due to issues with the mingw-w64 provided by msys2. At the very least you could consolidate that with the new information you're providing instead of having two separate somewhat redundant paragraphs.

If you feel up to it, could you add some more information clarifying that mingw-w64 only needs to be installed when building the `-gnu` targets and that when building the `-msvc` targets the user needs to have either VS 2013 or VS 2015 with the C++ tools installed instead. Also the `-msvc` targets need to have cmake installed to build LLVM. `-gnu` may need cmake too, I don't know whether we've transitioned to using cmake on all targets yet or just `-msvc` still.

**Comment by pitdicker** [comment]
It is just a detail, but I was confused whether I should choose the right msys flavor on Sourceforge or in the installer. Can you say "and choose the `threads=win32,exceptions=dwarf/seh` flavor in the installer"?

**Comment by steveklabnik** [comment]
Updated! I took some text from the install page.

**Comment by retep998** [comment]
Note that Msys2 still need to be installed, even when building the msvc targets, at least until @alexcrichton 's cargoified version of Rust is ready.

**Pull request merged**